Optimized EMA implementation

milesial opened the pull request:
This PR proposes an optimized way to do Exponential Moving Average (EMA), which is faster than the current way using `swa_utils.AveragedModel` described in https://pytorch.org/docs/stable/optim.html#custom-averaging-strategies.

This implementation is asynchronous, and is built as an optimizer wrapper so that the EMA weight update happens without any additional CPU/GPU sync, just after optimizer steps, and with limited code changes.

Example usage:
```
model = Model().to(device)
opt = torch.optim.Adam(model.parameters())

opt = EMAOptimizer(opt, device, 0.9999)

for epoch in range(epochs):
    training_loop(model, opt)

    regular_eval_accuracy = evaluate(model)

    with opt.swap_ema_weights():
        ema_eval_accuracy = evaluate(model)
```

Here are some benchmarks (time per iteration) on various torchvision models:

|model|this PR iteration time                      |swa_utils.AveragedModel iteration time| iteration speedup                                      |
|-----|-----------------------------|-----------------------|---------------------------------------------|
|     |                             |                       |                                             |
|regnet_x_1_6gf|62.73                        |67.998                 |1.08                                         |
|regnet_x_3_2gf|101.75                       |109.422                |1.08                                         |
|regnet_x_400mf|25.13                        |32.005                 |1.27                                         |
|regnet_x_800mf|33.01                        |37.466                 |1.13                                         |
|regnet_x_8gf|128.13                       |134.868                |1.05                                         |
|regnet_y_16gf|252.91                       |261.292                |1.03                                         |
|regnet_y_1_6gf|72.14                        |84.22                  |1.17                                         |
|regnet_y_3_2gf|99.99                        |109.296                |1.09                                         |
|regnet_y_400mf|29.53                        |36.506                 |1.24                                         |
|regnet_y_800mf|37.82                        |43.634                 |1.15                                         |
|regnet_y_8gf|196.63                       |203.317                |1.03                                         |
|resnet101|128.80                       |137.434                |1.07                                         |
|resnet152|182.85                       |196.498                |1.07                                         |
|resnet18|29.06                        |29.975                 |1.03                                         |
|resnet34|50.73                        |53.443                 |1.05                                         |
|resnet50|76.88                        |80.602                 |1.05                                         |
|resnext101_32x8d|277.29                       |280.759                |1.01                                         |
|resnext101_64x4d|269.56                       |281.052                |1.04                                         |
|resnext50_32x4d|100.73                       |101.102                |1.00                                         |
|shufflenet_v2_x0_5|10.56                        |15.419                 |1.46                                         |
|shufflenet_v2_x1_0|13.11                        |18.525                 |1.41                                         |
|shufflenet_v2_x1_5|18.05                        |23.132                 |1.28                                         |
|shufflenet_v2_x2_0|25.04                        |30.008                 |1.20                                         |
|squeezenet1_1|14.26                        |14.325                 |1.00                                         |
|swin_b|264.52                       |274.613                |1.04                                         |
|swin_s|180.66                       |188.914                |1.05                                         |
|swin_t|108.62                       |112.632                |1.04                                         |
|swin_v2_s|220.29                       |231.153                |1.05                                         |
|swin_v2_t|127.27                       |133.586                |1.05                                         |
|vgg11|95.52                        |103.714                |1.09                                         |
|vgg11_bn|106.49                       |120.711                |1.13                                         |
|vgg13|132.94                       |147.063                |1.11                                         |
|vgg13_bn|149.73                       |165.256                |1.10                                         |
|vgg16|158.19                       |172.865                |1.09                                         |
|vgg16_bn|177.04                       |192.888                |1.09                                         |
|vgg19|184.76                       |194.194                |1.05                                         |
|vgg19_bn|203.30                       |213.334                |1.05                                         |
|vit_b_16|217.31                       |219.748                |1.01                                         |
|vit_b_32|69.47                        |75.692                 |1.09                                         |
|vit_l_32|223.20                       |258.487                |1.16                                         |
|wide_resnet101_2|267.38                       |279.836                |1.05                                         |
|wide_resnet50_2|145.06                       |154.918                |1.07                                         |


You can see that in all cases it is faster than using `AveragedModel`. In fact in many cases, adding EMA does not add any overhead since the computation is hidden behind the usual iteration flow.

This is a similar implementation to the one currently in [NVIDIA NeMo](https://github.com/NVIDIA/NeMo).

If the team is interested in merging this, let me know and I'll add some documentation similar to `swa_utils` and tests.

Credits to @szmigacz for the implementation.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @mcarilli @ptrblck @leslie-fang-intel @EikanWang @soumith @voznesenskym @penguinwu @anijain2305 @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

milesial committed:
Optimized optimizer EMA

Skylion007 commented on torch/optim/ema_utils.py:
```diff
@@ -0,0 +1,198 @@
+        self.ema_params = ()
+
+    def all_parameters(self):
+        return (
```
Would `yield from` be better than return here?

albanD commented:
Hey!
Is there an issue discussing if we want to add this feature or not?
We usually prefer to have such discussion before you spend time writing down all the code.

milesial commented:
There is no issue open, I think we can discuss here? Or should I create one?
The code was already in NeMo and internally so it wasn't developed specifically for this PR.

vadimkantorov commented:
Related feature request: https://github.com/pytorch/pytorch/issues/71683

milesial commented:
> Also might be good to replace AveragedModel with this

`AveragedModel` works as a module wrapper while this works as an optimizer wrapper so this won't be a drop-in replacement, but we could provide alternative options to the `ema_update` function to do regular averaging for SWA for example if that's what you mean.

> And maybe to have an example of using this for explicit updates of BatchNorm stats?

This only tracks learnable parameters given to the optimizer and not buffers and BN stats, so there isn't anything special to do when BN are part of the model. These will be kept just as they are in the module.

> a functional form would also be nice for implementing a MeanTeacher optimzer as in the issue I linked

Yes this would be straight forward, reusing the `ema_update` function from this PR and just calling it in a separate CUDA stream

vadimkantorov commented:
> This only tracks learnable parameters given to the optimizer and not buffers and BN stats, so there isn't anything special to do when BN are part of the model. These will be kept just as they are in the module.

Well, I understand that. My idea is that we can have a separate set of running_stats parameters which are explicitly updated (either with a separate optimizer object or with a functional optimizer call). This makes it easy to protect batchnorm running params updates against infs/nans/large values which is currently possible but not very idiomatic. So yeah, currently these updates are taken care of by the modules forward call, but there's value in doing it explicitly in a nice way too.

Maybe `ema_update` could be adapted to be the functional form API contract, as is for all other optimizers? e.g. at https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L187

milesial commented:
> This makes it easy to protect batchnorm running params updates against infs/nans/large values which is currently possible but not very idiomatic.

This seems to me like you want to add some feature to batch norms layers themselves. I'm missing how this relates specifically to EMA at all. The goal of EMA is to not interact with running stats. Now if your running stats are broken because of infs even without EMA, then that's not in the scope of this PR. If the stats are not broken without EMA, then they won't be broken with EMA.

> Maybe ema_update could be adapted to be the functional form API contract, as is for all other optimizers

Can you provide the signature you're thinking about? `ema_update` takes some set of parameters and updates them inplace, which to me looks already the same as the `sgd` function for example.

> I think that it's more clean to keep this functionality as an optimizer wrapper, and have explicit calls for updates

Yes having the weight updates not called automatically after each optimizer step, and letting the user do it is of course an option. More code changes needed to enable EMA, but also more flexibility to call it conditionally for example

> asking the user to deal with creation of the twin model, and providing explicitly the target preallocated updatable by EMA param_groups and another optimizer that holds the online param_groups.

Yes we could offer the option to give to the class the parameters that need to be updated, and if these parameters are not given, then it would default to cloning the params internally.

vadimkantorov commented:
Yeah, the functional version would be versatile and nice to have, similar to torch.optim.functional.*

janeyx99 commented:
Thanks @milesial for wanting to contribute this! There are some rules regarding adding new features/functions to PyTorch Core here: https://github.com/pytorch/pytorch/wiki/Developer-FAQ#i-have-a-new-function-or-feature-id-like-to-add-to-pytorch-should-it-be-in-pytorch-core-or-a-library-like-torchvision.

In particular, I'm curious about the community engagement with EMA--could you point to evidence that this would be widely used (in terms of paper citations/code usages)? A quick GitHub search on my end reveals ~740 files importing EMA (https://github.com/search?q=%22ema+import+EMA%22&type=code), which might not justify the complexity of maintaining it in Core especially since we currently have no dedicated maintainer on this part of the codebase. Having this in NeMo may be the best course of action for now!

vadimkantorov commented:
On my side, I can add that MeanTeacher (essentially EMA) is used a lot in teacher-student models / self-supervised learning, and also as SWA averager (already implemented in pytorch, but maybe the existing API is a bit too specialized, and is module-oriented, not optimizer-oriented)

janeyx99 commented:
@vadimkantorov having a clean API for async updates as a generic util seems to be a different scope (beyond optimizers) than this particular PR and should be discussed more broadly in a separate issue if you would like.

vadimkantorov commented:
@janeyx99 Well, async updates are proposed in this optimizer in this PR, this was absent in other optimizers, so discussion about async updates popped up...

janeyx99 commented:
@vadimkantorov oh yea that's totally okay! Sorry I didn't mean to say it wasn't okay, but that further discussion should be in a different issue so we don't conflate this PR too much.

vadimkantorov commented:
@janeyx99 Well, if it is feasible to factor it out and without any loss of perf, maybe a better design decision is to not include it at all in the optimizer code itself

janeyx99 commented:
When you are talking about a generic util--do you mean for optimizers only? I was under the impression that you were suggesting the util as a generic torch API thing for more than just optimizers, which would entail a bigger discussion than should go on this thread.

vadimkantorov commented:
I don't have a very good idea :) I think if it be factored out, it can be used for anything (and maybe torch already has some types for futures / or python futures can be used), but we'd need to hear what @milesial thinks

vadimkantorov commented:
@milesial one more thing about async: maybe make it turned off by default (e.g. accept a `async = True` argument with `False` being default)? to avoid any surprises if the user does not know about this new async feature (not present in any other optimizers)

Wesley-Jzy commented:
@milesial Really great for me. May I know the reason why your EMA is faster?

janeyx99 commented:
@milesial Exciting that you'd be willing to step up to maintain this! I am working on establishing some recurring benchmarks for our optimizers for the next few weeks as a part of the effort to solidify torch.optim and will be able to dedicate more attention afterwards. In the meantime, we shouldn't hesitate to get this ball rolling. 

First of all, thanks for giving some context around EMA. To my understanding, this PR proposes two main pieces: using foreach_lerp as an EMA update and using async updates. I am curious about how you see these fitting in the PyTorch core offering, since we are less inclined to ad-hoc add another util vs continuing a storyline. 

1. For using foreach_lerp as a way to do speedy EMA--it would likely be better to figure out a design in the existing AveragedModel to allow for better extensibility e.g., incorporate allowing foreach type functions (like foreach_lerp) to pass into avg_fn (which currently does not work). We'd love to consider proposals here!
2. For async updates--we agree this is outside the scope of this specific RFC and so let's move any future discussion on this to a different issue.

vadimkantorov commented:
> 2\. For async updates--we agree this is outside the scope of this specific RFC and so let's move any future discussion on this to a different issue.

I just wanted to point out that this optimizer does do async updates already, and no other optimizer does that, which may come surprising to the users. But duly noted - let's discuss this somewhere else.

janeyx99 commented:
@vadimkantorov Yes, the current part under discussion has not moved far enough into implementation--we would like to have a design for incorporating performant EMA (ideally as a part of AveragedModel) and that design will likely look different than this current PR.

vadimkantorov commented on torch/optim/ema_utils.py:
```diff
@@ -0,0 +1,198 @@
+        """
+
+        def swap_tensors(tensor1, tensor2):
+            tmp = torch.empty_like(tensor1)
```
btw this might be done via simple swapping of storage() references? (and this would incur no copy / memory accesses?) I also wonder if there is an existing util function for this - seems a useful idiom.

If actual copies are important, there is also a related request on:
- https://github.com/pytorch/pytorch/issues/90539

milesial replied:
Swapping references seems a bit hacky to me. It might break code that relies on storage references and I'm not aware of any other pytorch op that manipulates storage references like that. But that's up for discussion.

This part is also not meant to be called very often (only before/after validation or saving weights), so the performance is not that critical. Once a `_foreach_copy` exists we could use it of course.

milesial commented:
> @milesial Really great for me. May I know the reason why your EMA is faster?

@Wesley-Jzy two things: use of the `_foreach` APIs and asynchonous work (CUDA stream or CPU thread)

> I am curious about how you see these fitting in the PyTorch core offering, since we are less inclined to ad-hoc add another util vs continuing a storyline.

@janeyx99 I was thinking that this is not really for me to tell. I proposed some idea/implementation but I'm not sure how this fits into your storyline since I'm not part of Meta :) So this is up to you.


> For using foreach_lerp as a way to do speedy EMA--it would likely be better to figure out a design in the existing AveragedModel to allow for better extensibility e.g., incorporate allowing foreach type functions (like foreach_lerp) to pass into avg_fn (which currently does not work). We'd love to consider proposals here!
> For async updates--we agree this is outside the scope of this specific RFC and so let's move any future discussion on this to a different issue.

So my understanding is that you areinterested in speeding up `AveragedModel` SWA/EMA using the `_foreach` APIs, but you're not interested yet in doing the EMA/SWA asynchronously using cuda streams or threads like it's done in the proposed implementation?


If that's the case, we could extend the existing `AverageModel` with a possibility to give an `avg_fn` alternative that works on pairs of sets of parameters instead of pairs of parameters. Let's say it's called `multi_avg_fn`

The signature would be changed from:
`AveragedModel(model, device=None, avg_fn=None, use_buffers=False)`
to:
`AveragedModel(model, device=None, avg_fn=None, multi_avg_fn=None, use_buffers=False)`

With only one of `avg_fn` and `multi_avg_fn` set. When `avg_fn` is set, we do the existing for loop:

```
for p_swa, p_model in zip(self_params, model_params):
    device = p_swa.device
    p_model_ = p_model.detach().to(device)
    if self.n_averaged == 0:
        p_swa.detach().copy_(p_model_)
    else:
        p_swa.detach().copy_(self.avg_fn(p_swa.detach(), p_model_,
                                         self.n_averaged.to(device)))
```

Else if `multi_avg_fn` is set we call it directly so that foreach APIs can be set.

```
multi_avg_fn(self_params, model_params)
```

We would propose some existing `multi_avg_fn` options to do SWA and EMA with `_foreach` APIs like the `ema_update` of the proposed implementation

```
def ema_update(ema_model_tuple, current_model_tuple, decay):
    torch._foreach_lerp_(ema_model_tuple, current_model_tuple, decay)
```

so that a user could do a faster ema with

`AveragedModel(model, multi_avg_fn=ema_update)`


Then we could talk later about calling the `multi_avg_fn` in a separate CUDA stream for maximum performance.

janeyx99 commented:
> @janeyx99 I was thinking that this is not really for me to tell. I proposed some idea/implementation but I'm not sure how this fits into your storyline since I'm not part of Meta :) So this is up to you.

Hahaha fair 😛 Since you're someone who is in touch with how people would use this feature, I wanted to hear your perspective and confirm that we would have a proposal that suffices for use cases! Thank you for the detailed proposal; I  agree it is a good way to make a faster EMA more widely available.

> Else if `multi_avg_fn` is set we call it directly so that foreach APIs can be used.
> 
> ```
> multi_avg_fn(self_params, model_params)
> ```

Getting into more details here: it looks like the current implementation handles when the self_params and model_params are on different devices. It'd be a good idea to add a check here + move the params to the same device before calling the custom function.

> We would propose some existing `multi_avg_fn` options to do SWA and EMA with `_foreach` APIs like the `ema_update` of the proposed implementation
> 
> ```
> def ema_update(ema_model_tuple, current_model_tuple, decay):
>     torch._foreach_lerp_(ema_model_tuple, current_model_tuple, decay)
> ```
> 
> so that a user could do a faster ema with
> 
> `AveragedModel(model, multi_avg_fn=ema_update)`

Yes! This would be great :D Would just have to make sure the `ema_update` function schema lines up correctly with the `multi_avg_fn` schema. I see that the ema_update takes in one extra arg, which we can pull out:
```
def ema_update(decay):
    def f(ema_model_tuple, current_model_tuple):
        return your_ema_update(ema_model_tuple, current_model_tuple, decay)
```
and the user would just put `AveragedModel(model, multi_avg_fn=ema_update(decay))`

> Then we could talk later about calling the `multi_avg_fn` in a separate CUDA stream for maximum performance.
👍

Another bigger picture concern (that I do not think is a blocker) is that the current AveragedModel is phrased to be SWA-specific and does not have a generic AveragedModel API. This would be a big change that can come after this implementation.

milesial committed:
Revert "Optimized optimizer EMA"

This reverts commit 3b1b82a7467db5722a2ae57a07eb8b49ce17363d.

milesial committed:
multi_avg_fn into AveragedModel

milesial commented:
@janeyx99 Sorry for the delay. I updated the code to add `multi_avg_fn` support into AveragedModel.
`multi_avg_fn` works on list of parameters that have the same device and dtype, and update the parameters inplace for maximum performance.

I added `get_ema_multi_avg_fn` and `get_swa_multi_avg_fn` as examples of using foreach functions for EMA and SWA.

Let me know if the logic sounds ok, and if so I'll add docs and tests.

janeyx99 commented on torch/optim/swa_utils.py:
```diff
@@ -125,14 +149,26 @@ def update_parameters(self, model):
+                p_swa.detach().copy_(self.avg_fn(p_swa.detach(), p_model,
                                                  self.n_averaged.to(device)))
+        elif self.n_averaged > 0 and self.multi_avg_fn is not None:
+            grouped_self = _group_tensors_by_device_and_dtype([self_param_detached])
+            grouped_model = _group_tensors_by_device_and_dtype([model_param_detached])
+            for device, dtype in grouped_self.keys():
+                self.multi_avg_fn(grouped_self[(device, dtype)][0], grouped_model[(device, dtype)][0], self.n_averaged.to(device))
```
```suggestion
            grouped_tensors = _group_tensors_by_device_and_dtype([self_param_detached, model_param_detached])
            for ((device, _), [self_params, model_params]) in grouped_self.items():
                self.multi_avg_fn(self_params, model_params, self.n_averaged.to(device))
```
This should be possible more better clarity

janeyx99 commented on torch/optim/swa_utils.py:
```diff
@@ -7,7 +7,26 @@
+def get_ema_multi_avg_fn(decay):
+    @torch.no_grad()
+    def ema_update(ema_model_tuple, current_model_tuple, _):
+        torch._foreach_lerp_(ema_model_tuple, current_model_tuple, decay)
```
Why are these tuples? Shouldn't they be param lists?

janeyx99 reviewed (commented):
Thanks for putting the work into this! The design looks good overall--I left some questions and feedback. Feel free to start working on tests and docs as well.

milesial committed:
Address review and add tests

milesial committed:
EMA documentation

milesial commented on torch/optim/swa_utils.py:
```diff
@@ -125,14 +149,26 @@ def update_parameters(self, model):
             device = p_swa.device
             p_model_ = p_model.detach().to(device)
+            model_param_detached.append(p_model_)
+            self_param_detached.append(p_swa.detach())
```
Fixed

milesial commented on torch/optim/swa_utils.py:
```diff
@@ -99,18 +123,18 @@ class AveragedModel(Module):
-            def avg_fn(averaged_model_parameter, model_parameter, num_averaged):
-                return averaged_model_parameter + \
-                    (model_parameter - averaged_model_parameter) / (num_averaged + 1)
+        if avg_fn is None and multi_avg_fn is None:
```
Done

milesial commented:
Sorry for the reviewer/labels spam, I rebased on master and force pushed and somehow everyone got added automatically.
@janeyx99 should be the only reviewer needed for now

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -254,13 +254,13 @@ algorithms.
-:mod:`torch.optim.swa_utils` implements Stochastic Weight Averaging (SWA). In particular,
-:class:`torch.optim.swa_utils.AveragedModel` class implements SWA models,
+:mod:`torch.optim.swa_utils` implements Stochastic Weight Averaging (SWA) and Exponential Moving Average (EMA). In particular,
+:class:`torch.optim.swa_utils.AveragedModel` class implements SWA and EMA models,
```
nit: either 
```suggestion
the :class:`torch.optim.swa_utils.AveragedModel` class implements SWA and EMA models,
```
or
```suggestion
:class:`torch.optim.swa_utils.AveragedModel` implements SWA and EMA models,
```

milesial replied:
Fixed

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -270,16 +270,44 @@ SWA has been proposed in `Averaging Weights Leads to Wider Optima and Better Gen
+EMA models are constructed by specifying the ``multi_avg_fn`` argument as follows:
 
-Here the model ``model`` can be an arbitrary :class:`torch.nn.Module` object. ``swa_model``
+>>> averaged_model = AveragedModel(model, multi_avg_fn=get_ema_multi_avg_fn(decay))
```
Should we also note that decay is something the user has to provide? Better yet, could it be made optional where if it's not provided, decay will default to a certain value?

milesial replied:
Added default value and formula

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -254,13 +254,13 @@ algorithms.
+:func:`torch.optim.swa_utils.update_bn` is a utility function used to update SWA/EMA batch
 normalization statistics at the end of training.
 
 SWA has been proposed in `Averaging Weights Leads to Wider Optima and Better Generalization`_.
```
Somewhere in the docs should motivate EMA as well.

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -270,16 +270,44 @@ SWA has been proposed in `Averaging Weights Leads to Wider Optima and Better Gen
+
+Here the model ``model`` can be an arbitrary :class:`torch.nn.Module` object. ``averaged_model``
 will keep track of the running averages of the parameters of the ``model``. To update these
 averages, you can use the :func:`update_parameters` function:
```
This is not the fault of this PR, but since we are here, I might as well point it out: it is unclear whether it is intended for the user to call update_parameters and when. From the usage below, it looks like it should be called after the optimizer.step() after a specified step number (swa_start) and it would be good to clarify this usage.

milesial replied:
Added a sentence to clarify

janeyx99 commented on docs/source/optim.rst:
```diff
@@ -270,16 +270,44 @@ SWA has been proposed in `Averaging Weights Leads to Wider Optima and Better Gen
+``avg_fn`` or ``multi_avg_fn`` parameters:
+
+- ``avg_fn`` allows defining a function operating on each parameter tuple (averaged parameter, model parameter) and should return the new averaged parameter.
+- ``multi_avg_fn`` allows defining more efficient operations acting on multiple (averaged parameter, model parameter) tuples at the same time, for example using the ``torch._foreach*`` functions. This function should update the averaged parameters in-place.
```
```suggestion
- ``multi_avg_fn`` allows defining more efficient operations acting on a tuple of parameter lists, (averaged parameter list, model parameter list), at the same time, for example using the ``torch._foreach*`` functions. This function must update the averaged parameters in-place.
```
It is more accurate to say the function operates on a tuple of lists instead of lists of tuples, and we should be clear that we expect inplace updates.

milesial replied:
Fixed

janeyx99 commented on torch/optim/swa_utils.py:
```diff
@@ -31,6 +55,11 @@ class AveragedModel(Module):
+            parameters inplace; the function must take in the current values of the
+            :class:`AveragedModel` parameters as a list, the current values of :attr:`model`
+            parameters as a list and the number of models already averaged; if None,
+            equally weighted average is used (default: None)
```
```suggestion
           an equally weighted average is used (default: None)
```
nit

milesial replied:
Fixed

janeyx99 commented on torch/optim/swa_utils.py:
```diff
@@ -121,23 +150,26 @@ def update_parameters(self, model):
                 p_swa.detach().copy_(p_model_)
+
+        if self.n_averaged > 0:
+            if self.multi_avg_fn is not None or self.avg_fn is None:
```
This would mean the default (if no avg_fn nor multi_avg_fn were provided) would call the path that would call torch.foreach, which is only defined for CUDA/CPU (and so tensors like mps tensors or any other custom device tensors would fail).

Could we have it so that the default only calls the multi_avg_fn if they're cuda tensors? This will require some more branching here.

janeyx99 replied:
looks good

janeyx99 commented:
Thanks @milesial for your work on this! Left another round of small comments and remember to fix lint!

milesial committed:
Fixes

milesial commented on torch/optim/swa_utils.py:
```diff
@@ -7,7 +7,31 @@
+
+def get_swa_multi_avg_fn():
+    @torch.no_grad()
+    def swa_update(ema_model_list, current_model_list, num_averaged):
```
Fixed

milesial commented on test/test_optim.py:
```diff
@@ -4026,30 +4034,39 @@ def _test_averaged_model(self, net_device, swa_device):
                 p.detach().add_(torch.randn_like(p))
-                p_avg += p.detach() / n_updates
+                if ema:
+                    p_avg += p.detach() * 0.999 ** (n_updates - i - 1) * (0.001 if i > 0 else 1.0)
```
Fixed

milesial commented on test/test_optim.py:
```diff
@@ -4010,13 +4010,21 @@ def _test_averaged_model(self, net_device, swa_device):
                 p.detach().add_(torch.randn_like(p))
-                p_avg += p.detach() / n_updates
+                if ema:
+                    p_avg += p.detach() * 0.999 ** (n_updates - i - 1) * (0.001 if i > 0 else 1.0)
```
Fixed

janeyx99 commented on test/test_optim.py:
```diff
@@ -4026,30 +4035,40 @@ def _test_averaged_model(self, net_device, swa_device):
 
-    def test_averaged_model_mixed_device(self):
+    @parametrize("ema", [True, False])
+    def test_averaged_model_mixed_device(self, ema):
```
Could this and _test_averaged_model  be refactored to call a common helper? There seems to be a bit of duplication.

milesial replied:
Done

janeyx99 commented on test/test_optim.py:
```diff
@@ -4119,19 +4143,24 @@ def avg_fn(p_avg, p, n_avg):
 
-    def test_averaged_model_exponential_buffers(self):
+    @parametrize("use_multi_avg_fn", [True, False])
+    def test_averaged_model_exponential_buffers(self, use_multi_avg_fn):
```
This test also looks very similar to test_averaged_model_exponential, could they abstract out the common areas?

milesial replied:
Done

milesial replied:
Merged them

janeyx99 commented on torch/optim/swa_utils.py:
```diff
@@ -30,7 +70,12 @@ class AveragedModel(Module):
+        multi_avg_fn (function, optional): the averaging function used to update
+            parameters inplace; the function must take in the current values of the
+            :class:`AveragedModel` parameters as a list, the current values of :attr:`model`
+            parameters as a list and the number of models already averaged; if None,
```
```suggestion
            parameters as a list, and the number of models already averaged; if None,
```
super super nit

milesial replied:
Fixed

janeyx99 reviewed (commented):
Thanks for the iteration :D This is very close--please fix lint and rebase!

janeyx99 commented on docs/source/compile/generated/torch.compile.rst:
```diff
@@ -0,0 +1,6 @@
+﻿torch.compile
```
why is this file necessary?

milesial replied:
Committed by mistake, removed

milesial committed:
Address review and lint

milesial commented on torch/optim/swa_utils.py:
```diff
@@ -121,28 +166,40 @@ def update_parameters(self, model):
             # keep the buffers in sync with the source model.
             for b_swa, b_model in zip(self.module.buffers(), model.buffers()):
-                b_swa.detach().copy_(b_model.detach().to(device))
+                b_swa.detach().copy_(b_model.detach().to(b_swa.device))
```
`device` was defined in some loop above so when buffers matched the device of the last parameter it worked

milesial commented on docs/source/optim.rst:
```diff
@@ -254,32 +254,73 @@ algorithms.
+Here the model ``model`` can be an arbitrary :class:`torch.nn.Module` object. ``averaged_model``
 will keep track of the running averages of the parameters of the ``model``. To update these
-averages, you can use the :func:`update_parameters` function:
+averages, you should use the :func:`update_parameters` function:
```
Fixed

milesial commented on docs/source/optim.rst:
```diff
@@ -254,32 +254,73 @@ algorithms.
+
 .. _`Averaging Weights Leads to Wider Optima and Better Generalization`: https://arxiv.org/abs/1803.05407
 
+.. _`Polyak averaging`: https://epubs.siam.org/doi/abs/10.1137/0330046
```
Yes polyak uses equal weights. Added some clarification and fixed the reference.

milesial committed:
Cleanup tests

janeyx99 commented on torch/optim/swa_utils.py:
```diff
@@ -7,7 +7,55 @@
+    return swa_update
 
 
 class AveragedModel(Module):
```
Do you want to also update the documentation here to include EMA in addition to SWA?

milesial replied:
Done

janeyx99 commented on torch/optim/swa_utils.py:
```diff
@@ -56,16 +109,15 @@ class AveragedModel(Module):
-        ...                 0.1 * averaged_model_parameter + 0.9 * model_parameter)
-        >>> swa_model = torch.optim.swa_utils.AveragedModel(model, avg_fn=ema_avg, use_buffers=True)
+        >>> swa_model = torch.optim.swa_utils.AveragedModel(model,
+        >>>             torch.optim.swa_utils.get_ema_multi_avg_fn(0.9), use_buffers=True)
```
Would be good to update the notes below to include multi_avg_fn + EMA

milesial replied:
Done

janeyx99 reviewed (commented):
I somehow do not see the updates for the docs in the first pytorchbot links--could you show a preview/screenshot of them?

In the process, I've also seen some more places that need updating. Excited to bring this over the finish line soon :D

janeyx99 commented:
Ah, the docs should preview correctly after https://github.com/pytorch/pytorch/pull/100022 is landed

milesial committed:
Update docstring

milesial commented:
I rebased to include https://github.com/pytorch/pytorch/pull/100022 so hopefully docs will be generated

janeyx99 reviewed (approved):
Looks good conditioning that the CI is green.

It would still be good to post a screenshot/link to the exact doc changes for future reference when they are built.

janeyx99 commented:
@pytorchbot merge

milesial commented:
Great, thanks! I'll work on a new MR for the async aspect of it, to make it even more efficient

janeyx99 commented:
@milesial thank you for bringing this PR through! Now let's just hope it makes its way through our internal diff train unscathed (which is not too high of a risk here). It would be interesting, if convenient, to measure the perf impact landed through this PR already as well.

For the async aspect, make sure you find someone that can review the change before implementing. Unfortunately, I currently do not have the expertise to validate and review the async component. This can be done by describing what you'd like to do in an issue.

netw0rkf10w commented:
@milesial This is a great PR, thanks a lot! Can it be applied to do weight averaging of the last `N` epochs? Even though this use case is less popular than EMA weight averaging, it is also very popular in NLP and speech. Example: it is used in the Attention Is All You Need paper (I guess no other examples are needed :P). Thanks in advance for your reply!

janeyx99 commented:
@netw0rkf10w I believe averaging across just the last N epochs can already be done (similar to the Putting it all together: SWA example). @milesial please feel free to correct me if I misunderstood!

milesial commented:
Yes you can call update_parameters when you want, including only during the last epochs to get SWA over the last epochs

netw0rkf10w commented:
@janeyx99 @milesial Thanks. Sorry I was not very clear in my comment. What I meant was to be able to do weight averaging and evaluate the averaged weight on-the-fly after very epoch (and not only at the end of training). For example, if I am at epoch 20, then the average is computed from epochs 11 to 20 (assuming `N=10`); and if I'm at epoch 21, then it is computed from epochs 12 to 21, etc. 
It seems that we'll have to keep in memory a deque of `N` weights...

milesial commented:
I don't think you can do that with the current API, `AveragedModel` only keeps one copy of the weights.

milesial commented:
As I said in an earlier comment, async will be in a future PR

milesial commented:
@netw0rkf10w I'll open a proposal issue next week for the async part of it

milesial commented on torch/optim/swa_utils.py:
```diff
@@ -7,30 +7,87 @@
+    @torch.no_grad()
+    def swa_update(averaged_param_list, current_param_list, num_averaged):
+        diffs = torch._foreach_sub(current_param_list, averaged_param_list)
+        torch._foreach_addcdiv_(averaged_param_list, diffs, [num_averaged + 1] * len(averaged_param_list))
```
True, thanks for catching that. I did it for EMA but forgot about SWA. Will submit a PR.

janeyx99 replied:
Ah yes, happy to review
